### ISSUE TITLE
systemd-resolved broken since commit "nss-resolve: varlink_call() set error_id only when r >= 0"

### ISSUE THREAD

**Issue by MartinX3** [systemd/systemd#18025]
**systemd version the issue has been seen with**
 247.2

**Used distribution**
 Arch Linux / EndeavourOS

**Linux kernel version used** (`uname -a`)
 5.4.84-1-lts

**CPU architecture issue was seen on**
 x86_64

**Expected behaviour you didn't see**
 Could resolve hostname

**Unexpected behaviour you saw**
 Couldn't resolve hostname

**Steps to reproduce the problem**
 Trying use DNS resolution

**Additional information:**
Happens since https://github.com/systemd/systemd/commit/3c229a9afb670025a3c1792bf5ac233b5a566ec5

[resolved.conf.txt](https://github.com/systemd/systemd/files/5719903/resolved.conf.txt)
[nsswitch.conf.txt](https://github.com/systemd/systemd/files/5719904/nsswitch.conf.txt)
[systemd-resolved.logs.txt](https://github.com/systemd/systemd/files/5719905/systemd-resolved.logs.txt)

**Comment by yuwata** [comment]
Hmm? The commit 3c229a9afb670025a3c1792bf5ac233b5a566ec5 does not change anything in resolved. Did you bisect and find the commit?

**Comment by eworm-de** [comment]
This was reported in Arch Linux bug tracker after we cherry-picked said commit...
https://bugs.archlinux.org/task/68981
https://github.com/archlinux/svntogit-packages/commit/8a4be64d166704be7db6a0cddba6e3aee93a51c8

**Comment by MartinX3** [comment]
And it seems that there some people fwce the same problem
https://github.com/NixOS/nixpkgs/pull/107067

**Comment by MartinX3** [comment]
Has this something to do with the problem?
```
Dez 20 11:30:57 systemd-resolved[364424]: Switching to DNS server my.router.ip.which.works.as.dns for interface xxxxxxx.
Dez 20 11:30:57 systemd-resolved[364424]: Transaction 35864 for <accounts.google.com IN A> on scope dns on xxxxxxx/* now complete with <attempts-max-reached> from none (unsigned).
Dez 20 11:30:57 systemd-resolved[364424]: Freeing transaction 38796.
Dez 20 11:30:57 systemd-resolved[364424]: varlink-20: Sending message: {"error":"io.systemd.Resolve.MaxAttemptsReached","parameters":{}}
Dez 20 11:30:57 systemd-resolved[364424]: varlink-20: varlink: changing state pending-method → idle-server
Dez 20 11:30:57 systemd-resolved[364424]: Freeing transaction 35864.
Dez 20 11:30:57 systemd-resolved[364424]: varlink-20: Got POLLHUP from socket.
Dez 20 11:30:57 systemd-resolved[364424]: varlink-20: varlink: changing state idle-server → pending-disconnect
Dez 20 11:30:57 systemd-resolved[364424]: varlink-20: varlink: changing state pending-disconnect → processing-disconnect
Dez 20 11:30:57 systemd-resolved[364424]: varlink-20: varlink: changing state processing-disconnect → disconnected
```

**Comment by MartinX3** [comment]
I was able to isolate the problem to the option `DNSOverTLS`
Since this commit the option breaks my dns resolution.
Without it, everything is fine.
Now I'll use only `DNSSEC` until the bug is fixed. :)

**Comment by keszybz** [comment]
https://github.com/systemd/systemd/issues/18025 is probably the same.

**Comment by mrc0mmand** [comment]
Just encountered this on latest Rawhide in the CentOS CI SELinux job when trying to upgrade to `systemd-247.3-3.fc35.x86_64.rpm`:

```
Feb 21 19:10:26 fedora systemd[1]: Reexecuting.
Feb 21 19:10:26 fedora systemd[1]: Assertion 'p->n_ref > 0' failed at src/shared/varlink.c:386, function varlink_unref(). Aborting.
Feb 21 19:10:26 fedora audit[41931]: ANOM_ABEND auid=4294967295 uid=0 gid=0 ses=4294
967295 subj=system_u:system_r:init_t:s0 pid=41931 comm="systemd" exe=2F7573722F6C69622F73797374656D642F73797374656D64202864656C6574656429 sig=6 res=1
Feb 21 19:10:26 fedora systemd-coredump[41932]: Due to PID 1 having crashed coredump collection will now be turned off.
Feb 21 19:10:26 fedora systemd-coredump[41932]: Process 41931 (systemd) of user 0 dumped core.
                                                
                                                Stack trace of thread 41931:
                                                #0  0x00007ff46546d58b kill (libc.so.6 + 0x3d58b)
                                                #1  0x0000562c096ca48e crash (/usr/lib/systemd/systemd (deleted) + 0x4848e)
                                                #2  0x00007ff4656134b0 __restore_rt (libpthread.so.0 + 0x134b0)
                                                #3  0x00007ff46546d292 raise (libc.so.6 + 0x3d292)
                                                #4  0x00007ff4654568a4 abort (libc.so.6 + 0x268a4)
                                                #5  0x00007ff4657da8d2 n/a (/usr/lib/systemd/libsystemd-shared-247.so (deleted) + 0x728d2)
Feb 21 19:10:26 fedora systemd[1]: Caught <ABRT>, dumped core as pid 41931.
Feb 21 19:10:26 fedora systemd[1]: Freezing execution.
```

**Comment by keszybz** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=1931034

**Comment by poettering** [comment]
I think the reports are probably unrelated issues. THe initial one looks like a DoT issue. The rh one looks to me like another instance of the issue fixed by 9807fdc1da8e037ddedfa4e2c6d2728b6e60051e.

The DoT issue is probably caused by enabling it on a DNS server that doesn't actually support DoT, and wher our fallback paths are wrong. Since those got heavily reworked in git main, I'd ather wait this one out, and see if this is reproducible wit current git main.

**Comment by poettering** [comment]
hmm let's drop this from the milestone. I think these are different issues, and would prefer if we can wait to see if any of these are reproducible with v248
